File: year2/week3/solution/week3.cpp
#include <iostream>
#include <unordered_set>

using Dictionary = std::unordered_set<std::string>;

bool is_valid(std::string letters, const Dictionary& d) {
    if (letters.size() == 0)
        return true;

    std::string tmp;
    for (std::size_t i = 0; i < letters.size(); ++i) {
        tmp += letters[i];
        if (d.find(tmp) != d.end()) {
            auto check = is_valid(letters.substr(i + 1), d);
            if (check) return true;
        }
    }
    return false;
}

int main() {
    Dictionary d;
    int n;
    std:: cin >> n;
    for (int i = 0; i < n; ++i) {
        std::string s;
        std::cin >> s;
        d.insert(s);
    }
    int m;
    std::cin >> m;
    for (int i = 0; i < m; ++i) {
        std::string sentence;
        std::cin >> sentence;
        std::cout << is_valid(sentence, d) << std::endl;
    }

    return 0;
}

File: year2/week3/solution/week3.py
def is_valid(sentence, dictionary):
    if not sentence:
        return True

    for i in range(len(sentence)):
        if sentence[:i + 1] in dictionary:
            check = is_valid(sentence[i + 1:], dictionary)
            if check:
                return True
    return False

def main():
    n = int(input().strip())
    dictionary = { input().strip() for i in range(n) }

    m = int(input().strip())
    print("\n".join(map(str, [int(is_valid(input().strip(), dictionary)) for i in range(m)])))

if __name__ == "__main__":
    main()
